### PR TITLE
Fix declareFieldLabels or declareLenses with DuplicateRecordFields

### LINKED ISSUE #323: declareFieldLabels generates bogus Labels in presence of DuplicateRecordFields
In my project I define a data type like this:
```Haskell
declareFieldLabels [d|data SortPosition t = SortPosition {
    mode :: R.Behavior t SortMode,
    list :: R.Behavior t [Task],
    before :: R.Behavior t (Maybe UUID)
}|]
```
in the expection, that this generates e.g. the optic label `#mode`.

That worked fine for me. It works for other data types and it worked for this type as well. But after some code reshuffling in my project which did move files but didn‘t touch the exact statements involved, the `#mode` label disappeared. Instead I can acces the optic like this
`labelOptic @"$sel:mode:SortPosition"`. I figured that out by using `-ddump-splice`. Which gives the following output:

```Haskell
    data SortPosition_a12ic t_a12ih
      = SortPosition_a12id (R.Behavior t_a12ih SortMode) (R.Behavior t_a12ih [Task]) (R.Behavior t_a12ih (Maybe UUID))
    instance (a_a12ii ~ R.Behavior t_a12ih SortMode,
              b_a12ij ~ R.Behavior t_a12ih SortMode) =>
             LabelOptic "$sel:mode:SortPosition" A_Lens (SortPosition_a12ic t_a12ih) (SortPosition_a12ic t_a12ih) a_a12ii b_a12ij where
      {-# INLINE labelOptic #-}
      labelOptic
        = lensVL
            (\ f_a12ik s_a12il
               -> case s_a12il of {
                    SortPosition_a12id x1_a12im x2_a12in x3_a12io
                      -> (fmap
                            (\ y_a12ip -> ((SortPosition_a12id y_a12ip) x2_a12in) x3_a12io))
                           (f_a12ik x1_a12im) })
    instance (a_a12iq ~ R.Behavior t_a12ih [Task],
              b_a12ir ~ R.Behavior t_a12ih [Task]) =>
             LabelOptic "$sel:list:SortPosition" A_Lens (SortPosition_a12ic t_a12ih) (SortPosition_a12ic t_a12ih) a_a12iq b_a12ir where
      {-# INLINE labelOptic #-}
      labelOptic
        = lensVL
            (\ f_a12is s_a12it
               -> case s_a12it of {
                    SortPosition_a12id x1_a12iu x2_a12iv x3_a12iw
                      -> (fmap
                            (\ y_a12ix -> ((SortPosition_a12id x1_a12iu) y_a12ix) x3_a12iw))
                           (f_a12is x2_a12iv) })
    instance (a_a12iy ~ R.Behavior t_a12ih (Maybe UUID),
              b_a12iz ~ R.Behavior t_a12ih (Maybe UUID)) =>
             LabelOptic "$sel:before:SortPosition" A_Lens (SortPosition_a12ic t_a12ih) (SortPosition_a12ic t_a12ih) a_a12iy b_a12iz where
      {-# INLINE labelOptic #-}
      labelOptic
        = lensVL
            (\ f_a12iA s_a12iB
               -> case s_a12iB of {
                    SortPosition_a12id x1_a12iC x2_a12iD x3_a12iE
                      -> (fmap
                            (\ y_a12iF -> ((SortPosition_a12id x1_a12iC) x2_a12iD) y_a12iF))
                           (f_a12iA x3_a12iE) })
```

Maybe someone has some insight into what could be wrong?
This happens with optics-th and optics 0.2.

### PR DESCRIPTION
**Pull request by adamgundry**
Fixes #323.

### CODE REVIEW

File: optics-th/src/Optics/TH/Internal/Product.hs [hunk 1]
```diff
@@ -95,9 +96,26 @@ makeFieldOpticsForDatatype rules info =
...
-  expandName :: [Name] -> Maybe Name -> [DefName]
-  expandName allFields = concatMap (_fieldToDef rules tyName allFields) . maybeToList
+-- | Map a (possibly missing) field's name to zero-to-many optic definitions
+expandName :: LensRules -> Name -> [D.ConstructorInfo] -> [Name] -> Maybe Name -> [DefName]
+expandName rules tyName cons allFields =
+    concatMap (_fieldToDef rules tyName allFields . stripSel) . maybeToList
+  where
+    -- When DuplicateRecordFields is enabled, reified datatypes contain
+    -- "mangled" field names that look like $sel:foo:MkT where foo is the field
+    -- name and MkT is the first data constructor.  We strip off the prefix and
```
Comment (arybczak) [inline optics-th/src/Optics/TH/Internal/Product.hs:106]: Is `MkT` always the first constructor, even in cases like `data T = MkT1 { name1 :: String } | MkT2 { name2 :: String }` `name2` will be mangled as `$sel:name2:mkT1`?

Comment (adamgundry) [reply optics-th/src/Optics/TH/Internal/Product.hs]: Yes, it's just the first constructor in the type, regardless of whether the field occurs or not. I've tweaked the test to check this, and cleaned up the code a bit. In particular it now checks that the suffix is the expected one, which should lead to easier to understand errors if GHC's behaviour should ever change.

File: optics-th/tests/Optics/TH/Tests/DuplicateRecordFields.hs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fforce-recomp #-}
```
Comment (arybczak) [inline optics-th/tests/Optics/TH/Tests/DuplicateRecordFields.hs:10]: Is this needed?

Comment (adamgundry) [reply optics-th/tests/Optics/TH/Tests/DuplicateRecordFields.hs]: Oops, I left that in by mistake. For some reason changing the TH code wasn't leading to recompilation when I was testing. Fixed.

**Pull request merged**